File: src/lib/repo.ts
import type { RepoSpec, ResolvedRepo } from "../types.js";

// Supported git hosts
const SUPPORTED_HOSTS = ["github.com", "gitlab.com", "bitbucket.org"];
const DEFAULT_HOST = "github.com";

/**
 * Parse a repository specification into host, owner and repo
 * Supports:
 * - github:owner/repo
 * - github:owner/repo@ref
 * - gitlab:owner/repo
 * - owner/repo (defaults to github.com)
 * - owner/repo@ref
 * - owner/repo#ref
 * - https://github.com/owner/repo
 * - https://github.com/owner/repo#ref
 * - https://gitlab.com/owner/repo
 * - https://github.com/owner/repo/tree/branch
 * - github.com/owner/repo
 */
export function parseRepoSpec(spec: string): RepoSpec | null {
  let input = spec.trim();
  let ref: string | undefined;
  let host: string = DEFAULT_HOST;

  // Handle shorthand prefixes: github:, gitlab:, bitbucket:
  if (input.startsWith("github:")) {
    host = "github.com";
    input = input.slice(7);
  } else if (input.startsWith("gitlab:")) {
    host = "gitlab.com";
    input = input.slice(7);
  } else if (input.startsWith("bitbucket:")) {
    host = "bitbucket.org";
    input = input.slice(10);
  }
  // Handle full URLs: https://github.com/owner/repo
  else if (input.match(/^https?:\/\//)) {
    try {
      const url = new URL(input);
      host = url.hostname;
      const pathParts = url.pathname.slice(1).split("/").filter(Boolean);

      if (pathParts.length < 2) {
        return null;
      }

      const owner = pathParts[0];
      let repo = pathParts[1];

      // Remove .git suffix if present
      if (repo.endsWith(".git")) {
        repo = repo.slice(0, -4);
      }

      const hashRef = url.hash ? decodeURIComponent(url.hash.slice(1)) : "";
      if (hashRef) {
        ref = hashRef;
      } else if (
        pathParts.length >= 4 &&
        (pathParts[2] === "tree" || pathParts[2] === "blob")
      ) {
        // Handle /tree/branch or /blob/branch URLs
        ref = pathParts[3];
      }

      return { host, owner, repo, ref };
    } catch {
      return null;
    }
  }
  // Handle host/owner/repo format: github.com/owner/repo
  else if (SUPPORTED_HOSTS.some((h) => input.startsWith(`${h}/`))) {
    const firstSlash = input.indexOf("/");
    host = input.slice(0, firstSlash);
    input = input.slice(firstSlash + 1);
  }
  // Not a repo format if it starts with @ (scoped npm package)
  else if (input.startsWith("@")) {
    return null;
  }
  // Must contain exactly one / to be a repo (owner/repo)
  else if (input.split("/").length !== 2) {
    return null;
  }

  // Extract ref from @ or # suffix
  // owner/repo@v1.0.0 or owner/repo#main
  const atIndex = input.indexOf("@");
  const hashIndex = input.indexOf("#");

  if (atIndex > 0) {
    ref = input.slice(atIndex + 1);
    input = input.slice(0, atIndex);
  } else if (hashIndex > 0) {
    ref = input.slice(hashIndex + 1);
    input = input.slice(0, hashIndex);
  }

  // Split into owner/repo
  const parts = input.split("/");
  if (parts.length !== 2 || !parts[0] || !parts[1]) {
    return null;
  }

  return {
    host,
    owner: parts[0],
    repo: parts[1],
    ref,
  };
}

/**
 * Check if a string looks like a repo spec rather than an npm package
 */
export function isRepoSpec(spec: string): boolean {
  const trimmed = spec.trim();

  // Explicit prefix (github:, gitlab:, bitbucket:)
  if (
    trimmed.startsWith("github:") ||
    trimmed.startsWith("gitlab:") ||
    trimmed.startsWith("bitbucket:")
  ) {
    return true;
  }

  // Git host URL (any host)
  if (trimmed.match(/^https?:\/\//)) {
    try {
      const url = new URL(trimmed);
      const parts = url.pathname.split("/").filter(Boolean);
      if (parts.length >= 2) {
        return true;
      }
    } catch {
      // Fall through
    }
  }

  // host/owner/repo format
  if (SUPPORTED_HOSTS.some((h) => trimmed.startsWith(`${h}/`))) {
    return true;
  }

  // Scoped npm packages start with @
  if (trimmed.startsWith("@")) {
    return false;
  }

  // owner/repo format (must have exactly one /)
  // But need to distinguish from things that aren't repos
  const parts = trimmed.split("/");
  if (parts.length === 2 && parts[0] && parts[1]) {
    // Extract the repo part (before any @ or #)
    const repoPart = parts[1].split("@")[0].split("#")[0];
    // Valid usernames and repos: alphanumeric, hyphens, underscores
    // Repos can also have dots
    const validOwner = /^[a-zA-Z0-9][a-zA-Z0-9-]*$/.test(parts[0]);
    const validRepo = /^[a-zA-Z0-9._-]+$/.test(repoPart);
    return validOwner && validRepo;
  }

  return false;
}

interface GitHubApiResponse {
  default_branch: string;
  clone_url: string;
  html_url: string;
}

interface GitLabApiResponse {
  default_branch: string;
  http_url_to_repo: string;
  web_url: string;
}

/**
 * Resolve a repo spec to full repository information using the appropriate API
 */
export async function resolveRepo(spec: RepoSpec): Promise<ResolvedRepo> {
  const { host, owner, repo, ref } = spec;

  if (host === "github.com") {
    return resolveGitHubRepo(host, owner, repo, ref);
  } else if (host === "gitlab.com") {
    return resolveGitLabRepo(host, owner, repo, ref);
  } else {
    // For unsupported hosts, assume default branch is "main"
    return {
      host,
      owner,
      repo,
      ref: ref || "main",
      repoUrl: `https://${host}/${owner}/${repo}`,
      displayName: `${host}/${owner}/${repo}`,
    };
  }
}

async function resolveGitHubRepo(
  host: string,
  owner: string,
  repo: string,
  ref?: string,
): Promise<ResolvedRepo> {
  const apiUrl = `https://api.github.com/repos/${owner}/${repo}`;

  const response = await fetch(apiUrl, {
    headers: {
      Accept: "application/vnd.github.v3+json",
      "User-Agent": "opensrc-cli",
    },
  });

  if (!response.ok) {
    if (response.status === 404) {
      throw new Error(
        `Repository "${owner}/${repo}" not found on GitHub. ` +
          `Make sure it exists and is public.`,
      );
    }
    if (response.status === 403) {
      throw new Error(
        `GitHub API rate limit exceeded. Try again later or authenticate.`,
      );
    }
    throw new Error(
      `Failed to fetch repository info: ${response.status} ${response.statusText}`,
    );
  }

  const data = (await response.json()) as GitHubApiResponse;
  const resolvedRef = ref || data.default_branch;

  return {
    host,
    owner,
    repo,
    ref: resolvedRef,
    repoUrl: `https://github.com/${owner}/${repo}`,
    displayName: `${host}/${owner}/${repo}`,
  };
}

async function resolveGitLabRepo(
  host: string,
  owner: string,
  repo: string,
  ref?: string,
): Promise<ResolvedRepo> {
  const projectPath = encodeURIComponent(`${owner}/${repo}`);
  const apiUrl = `https://gitlab.com/api/v4/projects/${projectPath}`;

  const response = await fetch(apiUrl, {
    headers: {
      "User-Agent": "opensrc-cli",
    },
  });

  if (!response.ok) {
    if (response.status === 404) {
      throw new Error(
        `Repository "${owner}/${repo}" not found on GitLab. ` +
          `Make sure it exists and is public.`,
      );
    }
    throw new Error(
      `Failed to fetch repository info: ${response.status} ${response.statusText}`,
    );
  }

  const data = (await response.json()) as GitLabApiResponse;
  const resolvedRef = ref || data.default_branch;

  return {
    host,
    owner,
    repo,
    ref: resolvedRef,
    repoUrl: `https://gitlab.com/${owner}/${repo}`,
    displayName: `${host}/${owner}/${repo}`,
  };
}

/**
 * Convert a repo display name back to host/owner/repo format
 */
export function displayNameToSpec(displayName: string): {
  host: string;
  owner: string;
  repo: string;
} | null {
  const parts = displayName.split("/");
  if (parts.length !== 3) {
    return null;
  }
  return { host: parts[0], owner: parts[1], repo: parts[2] };
}

/**
 * @deprecated Use displayNameToSpec instead
 */
export function displayNameToOwnerRepo(displayName: string): {
  owner: string;
  repo: string;
} | null {
  // Handle old format: owner--repo
  if (displayName.includes("--") && !displayName.includes("/")) {
    const parts = displayName.split("--");
    if (parts.length !== 2) {
      return null;
    }
    return { owner: parts[0], repo: parts[1] };
  }

  // Handle new format: host/owner/repo
  const spec = displayNameToSpec(displayName);
  if (!spec) {
    return null;
  }
  return { owner: spec.owner, repo: spec.repo };
}

File: packages/opensrc/cli/src/core/registries/repo.rs
use std::sync::LazyLock;

use serde::Deserialize;

static RE_URL: LazyLock<regex::Regex> = LazyLock::new(|| {
    regex::Regex::new(r"^https?://(github\.com|gitlab\.com|bitbucket\.org)/").unwrap()
});
static RE_OWNER: LazyLock<regex::Regex> =
    LazyLock::new(|| regex::Regex::new(r"^[a-zA-Z0-9][a-zA-Z0-9-]*$").unwrap());
static RE_REPO: LazyLock<regex::Regex> =
    LazyLock::new(|| regex::Regex::new(r"^[a-zA-Z0-9._-]+$").unwrap());

const SUPPORTED_HOSTS: &[&str] = &["github.com", "gitlab.com", "bitbucket.org"];
const DEFAULT_HOST: &str = "github.com";

#[derive(Debug, Clone)]
pub struct RepoSpec {
    pub host: String,
    pub owner: String,
    pub repo: String,
    pub git_ref: Option<String>,
}

#[derive(Debug, Clone)]
pub struct ResolvedRepo {
    pub git_ref: String,
    pub repo_url: String,
    pub display_name: String,
}

pub fn parse_repo_spec(spec: &str) -> Option<RepoSpec> {
    let input = spec.trim();
    let mut remaining = input.to_string();
    let mut git_ref: Option<String> = None;
    let mut host = DEFAULT_HOST.to_string();

    // Handle shorthand prefixes
    if remaining.starts_with("github:") {
        host = "github.com".to_string();
        remaining = remaining[7..].to_string();
    } else if remaining.starts_with("gitlab:") {
        host = "gitlab.com".to_string();
        remaining = remaining[7..].to_string();
    } else if remaining.starts_with("bitbucket:") {
        host = "bitbucket.org".to_string();
        remaining = remaining[10..].to_string();
    } else if remaining.starts_with("http://") || remaining.starts_with("https://") {
        // Full URL
        let url = match url::Url::parse(&remaining) {
            Ok(u) => u,
            Err(_) => return None,
        };
        host = url.host_str()?.to_string();
        let path_parts: Vec<&str> = url
            .path()
            .trim_start_matches('/')
            .split('/')
            .filter(|s| !s.is_empty())
            .collect();

        if path_parts.len() < 2 {
            return None;
        }

        let owner = path_parts[0].to_string();
        let mut repo = path_parts[1].to_string();
        if repo.ends_with(".git") {
            repo = repo[..repo.len() - 4].to_string();
        }

        if path_parts.len() >= 4 && (path_parts[2] == "tree" || path_parts[2] == "blob") {
            git_ref = Some(path_parts[3].to_string());
        }

        return Some(RepoSpec {
            host,
            owner,
            repo,
            git_ref,
        });
    } else if SUPPORTED_HOSTS
        .iter()
        .any(|h| remaining.starts_with(&format!("{h}/")))
    {
        if let Some(idx) = remaining.find('/') {
            host = remaining[..idx].to_string();
            remaining = remaining[idx + 1..].to_string();
        }
    } else if remaining.starts_with('@') || remaining.split('/').count() != 2 {
        return None;
    }

    // Extract ref from @ or #
    if let Some(at_idx) = remaining.find('@') {
        if at_idx > 0 {
            git_ref = Some(remaining[at_idx + 1..].to_string());
            remaining = remaining[..at_idx].to_string();
        }
    } else if let Some(hash_idx) = remaining.find('#') {
        if hash_idx > 0 {
            git_ref = Some(remaining[hash_idx + 1..].to_string());
            remaining = remaining[..hash_idx].to_string();
        }
    }

    let parts: Vec<&str> = remaining.split('/').collect();
    if parts.len() != 2 || parts[0].is_empty() || parts[1].is_empty() {
        return None;
    }

    Some(RepoSpec {
        host,
        owner: parts[0].to_string(),
        repo: parts[1].to_string(),
        git_ref,
    })
}

pub fn is_repo_spec(spec: &str) -> bool {
    let trimmed = spec.trim();

    if trimmed.starts_with("github:")
        || trimmed.starts_with("gitlab:")
        || trimmed.starts_with("bitbucket:")
    {
        return true;
    }

    if RE_URL.is_match(trimmed) {
        return true;
    }

    if SUPPORTED_HOSTS
        .iter()
        .any(|h| trimmed.starts_with(&format!("{h}/")))
    {
        return true;
    }

    if trimmed.starts_with('@') {
        return false;
    }

    let parts: Vec<&str> = trimmed.split('/').collect();
    if parts.len() == 2 && !parts[0].is_empty() && !parts[1].is_empty() {
        let repo_part = parts[1]
            .split('@')
            .next()
            .unwrap_or("")
            .split('#')
            .next()
            .unwrap_or("");
        return RE_OWNER.is_match(parts[0]) && RE_REPO.is_match(repo_part);
    }

    false
}

#[derive(Deserialize)]
struct GitHubApiResponse {
    default_branch: String,
}

#[derive(Deserialize)]
struct GitLabApiResponse {
    default_branch: Option<String>,
}

#[derive(Deserialize)]
struct BitbucketMainBranch {
    name: Option<String>,
}

#[derive(Deserialize)]
struct BitbucketApiResponse {
    mainbranch: Option<BitbucketMainBranch>,
}

pub fn resolve_repo(spec: &RepoSpec) -> Result<ResolvedRepo, Box<dyn std::error::Error>> {
    match spec.host.as_str() {
        "github.com" => resolve_github(spec),
        "gitlab.com" => resolve_gitlab(spec),
        "bitbucket.org" => resolve_bitbucket(spec),
        _ => Ok(ResolvedRepo {
            git_ref: spec.git_ref.clone().unwrap_or_else(|| "main".to_string()),
            repo_url: format!("https://{}/{}/{}", spec.host, spec.owner, spec.repo),
            display_name: format!("{}/{}/{}", spec.host, spec.owner, spec.repo),
        }),
    }
}

fn resolve_github(spec: &RepoSpec) -> Result<ResolvedRepo, Box<dyn std::error::Error>> {
    let url = format!("https://api.github.com/repos/{}/{}", spec.owner, spec.repo);

    let client = super::http_client();
    let mut req = client
        .get(&url)
        .header("Accept", "application/vnd.github.v3+json");

    if let Some(token) = super::github_token() {
        req = req.header("Authorization", format!("Bearer {token}"));
    }

    let resp = req.send()?;

    if resp.status() == reqwest::StatusCode::NOT_FOUND {
        let hint = if super::github_token().is_none() {
            " If this is a private repo, set GITHUB_TOKEN."
        } else {
            " Your token may lack access to this repository."
        };
        return Err(format!(
            "Repository \"{}/{}\" not found on GitHub.{hint}",
            spec.owner, spec.repo
        )
        .into());
    }
    if resp.status() == reqwest::StatusCode::FORBIDDEN {
        return Err("GitHub API rate limit exceeded. Try again later or set GITHUB_TOKEN.".into());
    }
    if !resp.status().is_success() {
        return Err(format!("Failed to fetch repository info: {}", resp.status()).into());
    }

    let data: GitHubApiResponse = resp.json()?;
    let resolved_ref = spec.git_ref.clone().unwrap_or(data.default_branch);

    Ok(ResolvedRepo {
        git_ref: resolved_ref,
        repo_url: format!("https://github.com/{}/{}", spec.owner, spec.repo),
        display_name: format!("{}/{}/{}", spec.host, spec.owner, spec.repo),
    })
}

fn resolve_gitlab(spec: &RepoSpec) -> Result<ResolvedRepo, Box<dyn std::error::Error>> {
    let project_path = format!("{}/{}", spec.owner, spec.repo);
    let encoded = urlencoding::encode(&project_path);
    let url = format!("https://gitlab.com/api/v4/projects/{encoded}");

    let client = super::http_client();
    let mut req = client.get(&url);

    if let Some(token) = super::gitlab_token() {
        req = req.header("PRIVATE-TOKEN", &token);
    }

    let resp = req.send()?;

    if resp.status() == reqwest::StatusCode::NOT_FOUND {
        let hint = if super::gitlab_token().is_none() {
            " If this is a private repo, set GITLAB_TOKEN."
        } else {
            " Your token may lack access to this repository."
        };
        return Err(format!(
            "Repository \"{}/{}\" not found on GitLab.{hint}",
            spec.owner, spec.repo
        )
        .into());
    }
    if !resp.status().is_success() {
        return Err(format!("Failed to fetch repository info: {}", resp.status()).into());
    }

    let data: GitLabApiResponse = resp.json()?;
    let resolved_ref = spec
        .git_ref
        .clone()
        .unwrap_or_else(|| data.default_branch.unwrap_or_else(|| "main".to_string()));

    Ok(ResolvedRepo {
        git_ref: resolved_ref,
        repo_url: format!("https://gitlab.com/{}/{}", spec.owner, spec.repo),
        display_name: format!("{}/{}/{}", spec.host, spec.owner, spec.repo),
    })
}

fn resolve_bitbucket(spec: &RepoSpec) -> Result<ResolvedRepo, Box<dyn std::error::Error>> {
    let url = format!(
        "https://api.bitbucket.org/2.0/repositories/{}/{}",
        spec.owner, spec.repo
    );

    let client = super::http_client();
    let mut req = client.get(&url);

    if let Some(token) = super::bitbucket_token() {
        req = req.header("Authorization", format!("Bearer {token}"));
    }

    let resp = req.send()?;

    if resp.status() == reqwest::StatusCode::NOT_FOUND {
        let hint = if super::bitbucket_token().is_none() {
            " If this is a private repo, set BITBUCKET_TOKEN."
        } else {
            " Your token may lack access to this repository."
        };
        return Err(format!(
            "Repository \"{}/{}\" not found on Bitbucket.{hint}",
            spec.owner, spec.repo
        )
        .into());
    }
    if resp.status() == reqwest::StatusCode::UNAUTHORIZED
        || resp.status() == reqwest::StatusCode::FORBIDDEN
    {
        let hint = if super::bitbucket_token().is_none() {
            " If this is a private repo, set BITBUCKET_TOKEN."
        } else {
            " Your token may lack access to this repository."
        };
        return Err(format!(
            "Access denied to Bitbucket repository \"{}/{}\".{hint}",
            spec.owner, spec.repo
        )
        .into());
    }
    if !resp.status().is_success() {
        return Err(format!("Failed to fetch repository info: {}", resp.status()).into());
    }

    let data: BitbucketApiResponse = resp.json()?;
    let resolved_ref = spec.git_ref.clone().unwrap_or_else(|| {
        data.mainbranch
            .and_then(|b| b.name)
            .unwrap_or_else(|| "main".to_string())
    });

    Ok(ResolvedRepo {
        git_ref: resolved_ref,
        repo_url: format!("https://bitbucket.org/{}/{}", spec.owner, spec.repo),
        display_name: format!("{}/{}/{}", spec.host, spec.owner, spec.repo),
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_repo_spec_simple() {
        let spec = parse_repo_spec("vercel/next.js").unwrap();
        assert_eq!(spec.host, "github.com");
        assert_eq!(spec.owner, "vercel");
        assert_eq!(spec.repo, "next.js");
        assert_eq!(spec.git_ref, None);
    }

    #[test]
    fn test_parse_repo_spec_with_ref() {
        let spec = parse_repo_spec("vercel/next.js@canary").unwrap();
        assert_eq!(spec.owner, "vercel");
        assert_eq!(spec.repo, "next.js");
        assert_eq!(spec.git_ref, Some("canary".into()));
    }

    #[test]
    fn test_parse_repo_spec_github_prefix() {
        let spec = parse_repo_spec("github:vercel/next.js").unwrap();
        assert_eq!(spec.host, "github.com");
        assert_eq!(spec.owner, "vercel");
    }

    #[test]
    fn test_parse_repo_spec_full_url() {
        let spec = parse_repo_spec("https://github.com/vercel/next.js/tree/canary").unwrap();
        assert_eq!(spec.host, "github.com");
        assert_eq!(spec.owner, "vercel");
        assert_eq!(spec.repo, "next.js");
        assert_eq!(spec.git_ref, Some("canary".into()));
    }

    #[test]
    fn test_parse_repo_spec_bitbucket_prefix() {
        let spec = parse_repo_spec("bitbucket:atlassian/python-bitbucket").unwrap();
        assert_eq!(spec.host, "bitbucket.org");
        assert_eq!(spec.owner, "atlassian");
        assert_eq!(spec.repo, "python-bitbucket");
        assert_eq!(spec.git_ref, None);
    }

    #[test]
    fn test_parse_repo_spec_bitbucket_prefix_with_ref() {
        let spec = parse_repo_spec("bitbucket:atlassian/python-bitbucket@master").unwrap();
        assert_eq!(spec.host, "bitbucket.org");
        assert_eq!(spec.owner, "atlassian");
        assert_eq!(spec.repo, "python-bitbucket");
        assert_eq!(spec.git_ref, Some("master".into()));
    }

    #[test]
    fn test_parse_repo_spec_bitbucket_url() {
        let spec = parse_repo_spec("https://bitbucket.org/atlassian/python-bitbucket").unwrap();
        assert_eq!(spec.host, "bitbucket.org");
        assert_eq!(spec.owner, "atlassian");
        assert_eq!(spec.repo, "python-bitbucket");
        assert_eq!(spec.git_ref, None);
    }

    #[test]
    fn test_is_repo_spec() {
        assert!(is_repo_spec("vercel/next.js"));
        assert!(is_repo_spec("github:vercel/next.js"));
        assert!(is_repo_spec("https://github.com/vercel/next.js"));
        assert!(is_repo_spec("bitbucket:atlassian/python-bitbucket"));
        assert!(is_repo_spec(
            "https://bitbucket.org/atlassian/python-bitbucket"
        ));
        assert!(!is_repo_spec("@babel/core"));
        assert!(!is_repo_spec("zod"));
    }
}
